tests/lib/hardware/libvirt.py

```diff
@@ -0,0 +1,300 @@
+# Copyright (c) 2019 SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# The Hardware module should take care of the operating system abstraction
+# through images.
+# libcloud will provide a common set of cloud-agnostic objects such as Node[s]
+# We might extend the Node object to have an easy way to run arbitrary commands
+# on the node such as Node.execute().
+# There will be a challenge where those arbitrary commands differ between OS's;
+# this is an abstraction that is not yet well figured out, but will likely
+# take the form of cloud-init or similar bringing the target node to an
+# expected state.
+
+import os
+import shutil
+import subprocess
+import time
+import tempfile
+import threading
+import datetime
+import logging
+import libvirt
+import uuid
+import paramiko
+import socket
+from typing import List
+from xml.dom import minidom
+
+from tests.lib.distro import get_distro
+from tests.lib.hardware.hardware_base import HardwareBase
+from tests.lib.hardware.node_base import NodeBase, NodeRole
+from tests import config
+
+logger = logging.getLogger(__name__)
+
+
+class Node(NodeBase):
+    def __init__(self, name, role, tags, conn, network, disk_number, memory,
+                 ssh_public_key, ssh_private_key):
+        super().__init__(name, role, tags)
+        self._conn = conn
+        self._network = network
+        self._disk_number = disk_number,
+        self._memory = memory * 1024 * 1024
+        self._ssh_public_key = ssh_public_key
+        self._ssh_private_key = ssh_private_key
+        self._snap_img_path = os.path.join(
+            os.path.dirname(config.PROVIDER_LIBVIRT_IMAGE),
+            f"{self.name}-snapshot.qcow2")
+        self._cloud_init_seed_path = os.path.join(os.path.dirname(
+            config.PROVIDER_LIBVIRT_IMAGE), f"{self.name}-cloud-init-seed.img")
```

Comment:
We should create these `self.*_path`'s in the working dir. Otherwise a user may provide a libvirt image in their `$HOME/Downloads/` folder and have other stuff created there unexpectedly. Additionally this will make it easier for us to provide an image as a URL rather than file (that could then be cached in the working_dir).

Reply:
Fixed in #55